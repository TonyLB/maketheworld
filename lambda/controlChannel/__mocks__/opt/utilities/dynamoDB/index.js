const dbMock = () => ({
    getItem: jest.fn(),
    batchGetItem: jest.fn(),
    query: jest.fn(),
    update: jest.fn(),
    putItem: jest.fn(),
    getItem: jest.fn()
})

export const assetDB = dbMock()
export const ephemeraDB = dbMock()

export const batchWriteDispatcher = jest.fn()
export const mergeIntoDataRange = jest.fn()
