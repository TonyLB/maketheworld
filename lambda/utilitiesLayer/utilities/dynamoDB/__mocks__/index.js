import { jest } from '@jest/globals'

const dbMock = () => ({
    getItem: jest.fn(),
    batchGetItem: jest.fn(),
    query: jest.fn(),
    update: jest.fn(),
    putItem: jest.fn(),
    getItem: jest.fn()
})

export const assetDB = dbMock()
export const ephemeraDB = dbMock()

export const mergeIntoDataRange = jest.fn()
export const batchWriteDispatcher = jest.fn()
