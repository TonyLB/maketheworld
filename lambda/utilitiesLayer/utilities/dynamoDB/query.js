import {
    QueryCommand,
} from "@aws-sdk/client-dynamodb"
import { marshall, unmarshall } from "@aws-sdk/util-dynamodb"

import { asyncSuppressExceptions } from '../errors.js'

const { TABLE_PREFIX } = process.env;
const ephemeraTable = `${TABLE_PREFIX}_ephemera`
const assetsTable = `${TABLE_PREFIX}_assets`
const messageTable = `${TABLE_PREFIX}_messages`
const deltaTable = `${TABLE_PREFIX}_message_delta`

const extractKeyInfo = (table, indexName, {
    AssetId,
    EphemeraId,
    MessageId,
    DataCategory,
    player,
    scopedId
}) => {
    if (indexName) {
        switch(indexName) {
            case 'DataCategoryIndex':
                return {
                    KeyConditionExpression: 'DataCategory = :keyId',
                    keyId: DataCategory
                }
            case 'PlayerIndex':
                return {
                    KeyConditionExpression: 'player = :keyId',
                    keyId: player
                }
            case 'ScopedIdIndex':
                return {
                    KeyConditionExpression: 'scopedId = :keyId',
                    keyId: scopedId
                }
        }    
    }
    else {
        switch(table) {
            case messageTable:
                return {
                    KeyConditionExpression: 'MessageId = :keyId',
                    keyId: MessageId,
                }
            case ephemeraTable:
                return {
                    KeyConditionExpression: 'EphemeraId = :keyId',
                    keyId: EphemeraId,
                }
            default:
                return {
                    KeyConditionExpression: 'AssetId = :keyId',
                    keyId: AssetId,
                }
        }
    }
}

export const abstractQueryExtended = (dbClient, table) => async (props) => {
    const {
        IndexName,
        ProjectionFields: passedProjectionFields,
        KeyConditionExpression: extraExpression,
        ExpressionAttributeNames,
        ExpressionAttributeValues,
        FilterExpression
    } = props
    return await asyncSuppressExceptions(async () => {
        const ProjectionFields = passedProjectionFields || 
            (IndexName
                ? [
                    table === assetsTable && 'AssetId',
                    table === ephemeraTable && 'EphemeraId',
                    table === messageTable && 'MessageId'
                ].filter((value) => (value)) 
                : ['DataCategory'])
        const { KeyConditionExpression: baseExpression, keyId } = extractKeyInfo(
            table,
            IndexName,
            props
        )
        const KeyConditionExpression = extraExpression
            ? `${baseExpression} AND ${extraExpression}`
            : baseExpression
        const { Items = [] } = await dbClient.send(new QueryCommand({
            TableName: table,
            KeyConditionExpression,
            IndexName,
            ExpressionAttributeValues: marshall({
                ':keyId': keyId,
                ...(ExpressionAttributeValues || {})
            }),
            ProjectionExpression: ProjectionFields.join(', '),
            ExpressionAttributeNames,
            FilterExpression
        }))
        return {
            Items: Items.map(unmarshall)
        }
    }, () => ({ Items: [] }))
}

export const abstractQuery = (dbClient, table) => async (props) => {
    const { Items } = await abstractQueryExtended(dbClient, table)(props)
    return Items
}