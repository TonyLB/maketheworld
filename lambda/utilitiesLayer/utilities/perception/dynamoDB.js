import { ephemeraDB } from '../dynamoDB/index.js'
import { AssetKey } from '../types.js'
import { splitType } from '../types.js'

export const getItemMeta = async (items) => {
    const getSingleMeta = async (EphemeraId) => {
        switch(splitType(EphemeraId)[0]) {
            case 'CHARACTERINPLAY':
                const characterItem = await ephemeraDB.getItem({
                    EphemeraId,
                    DataCategory: 'Meta::Character',
                    ProjectionFields: ['#name', 'fileURL'],
                    ExpressionAttributeNames: {
                        '#name': 'Name'
                    }
                })
                return { [EphemeraId]: [{
                    DataCategory: 'Meta::Character',
                    ...characterItem
                }] }
            default:
                const metaItems = await ephemeraDB.query({
                    EphemeraId,
                    ProjectionFields: ['DataCategory', 'appearances', '#name', 'activeCharacters'],
                    ExpressionAttributeNames: {
                        '#name': 'name'
                    }
                })
                return { [EphemeraId]: metaItems }        
        }
    }
    const allRooms = await Promise.all(items.map(getSingleMeta))
    return Object.assign({}, ...allRooms)
}

export const getGlobalAssets = async (priorGlobalAssets) => {
    if (priorGlobalAssets !== undefined) {
        return priorGlobalAssets
    }
    const { assets = [] } = await ephemeraDB.getItem({
        EphemeraId: 'Global',
        DataCategory: 'Assets',
        ProjectionFields: ['assets']
    })
    return assets
}

export const getCharacterAssets = async (characters, priorCharacterAssets = {}) => {
    //
    // TODO: Add RoomId to projection fields, and then use that to populate the
    // RoomId for any links rendered for that character
    //
    const getSingleCharacterAssets = async (characterId) => {
        const { assets = [], RoomId } = await ephemeraDB.getItem({
            EphemeraId: `CHARACTERINPLAY#${characterId}`,
            DataCategory: 'Meta::Character',
            ProjectionFields: ['assets', 'RoomId']
        })
        return { [characterId]: { assets, RoomId } }
    }
    const neededCharacters = await Promise.all(
        characters
            .filter((character) => (!(character in priorCharacterAssets)))
            .map(getSingleCharacterAssets)
    )
    return Object.assign(priorCharacterAssets, ...neededCharacters)
}

export const getStateByAsset = async (assets, existingStatesByAsset = {}) => {
    const getSingleState = async (assetId, existingStatesByAsset = {}) => {
        if (existingStatesByAsset[assetId]) {
            return {
                [assetId]: {
                    ...existingStatesByAsset[assetId],
                    State: Object.entries(existingStatesByAsset[assetId]?.State || {})
                        .reduce((previous, [key, { value }]) => ({ ...previous, [key]: value }), {}),
                }
            }
        }
        const { State = {}, mapCache = {} } = await ephemeraDB.getItem({
            EphemeraId: AssetKey(assetId),
            DataCategory: 'Meta::Asset',
            ProjectionFields: ['#state', 'mapCache'],
            ExpressionAttributeNames: {
                '#state': 'State'
            }
        })
        return {
            [assetId]: {
                State: Object.entries(State).reduce((previous, [key, { value }]) => ({ ...previous, [key]: value }), {}),
                mapCache
            }
        }
    }
    const allStates = await Promise.all(assets.map((assetId) => (getSingleState(assetId, existingStatesByAsset))))
    return Object.assign({}, ...allStates)
}

export const getNormalForm = async (assetId, existingNormalFormsByAsset = {}) => {
    if (existingNormalFormsByAsset[assetId]) {
        return existingNormalFormsByAsset[assetId]
    }
    const { normalForm = {} } = await ephemeraDB.getItem({
        EphemeraId: AssetKey(assetId),
        DataCategory: 'Meta::AssetNormalized',
        ProjectionFields: ['normalForm']
    })
    return normalForm
}