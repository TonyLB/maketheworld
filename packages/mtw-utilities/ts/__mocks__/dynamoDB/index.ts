const dbMock = () => ({
    getItem: jest.fn(),
    batchGetItem: jest.fn(),
    query: jest.fn(),
    update: jest.fn(),
    putItem: jest.fn(),
    deleteItem: jest.fn()
})

export const legacyAssetDB = dbMock()
export const ephemeraDB = dbMock()
export const assetDB = dbMock()
export const connectionDB = dbMock()

export const batchWriteDispatcher = jest.fn()
export const mergeIntoDataRange = jest.fn()

export const publishMessage = jest.fn()
export const messageDeltaQuery = jest.fn()

export const exponentialBackoffWrapper = jest.fn((testFunc) => { testFunc() })
export const multiTableTransactWrite = jest.fn()
