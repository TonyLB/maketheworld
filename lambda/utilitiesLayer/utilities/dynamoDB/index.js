import {
    DynamoDBClient,
    GetItemCommand,
    UpdateItemCommand,
    PutItemCommand,
    DeleteItemCommand,
    QueryCommand,
    BatchWriteItemCommand,
    BatchGetItemCommand,
    ScanCommand
} from "@aws-sdk/client-dynamodb"
import { marshall, unmarshall } from "@aws-sdk/util-dynamodb"

import { asyncSuppressExceptions } from '../errors.js'
import { abstractQuery } from './query.js'

const { TABLE_PREFIX } = process.env;
const ephemeraTable = `${TABLE_PREFIX}_ephemera`
const assetsTable = `${TABLE_PREFIX}_assets`
const messageTable = `${TABLE_PREFIX}_messages`
const deltaTable = `${TABLE_PREFIX}_message_delta`

const params = { region: process.env.AWS_REGION }
const dbClient = new DynamoDBClient(params)

export const batchWriteDispatcher = ({ table, items }) => {
    const groupBatches = items
        .reduce((({ current, requestLists }, item) => {
            if (current.length > 19) {
                return {
                    requestLists: [ ...requestLists, current ],
                    current: [item]
                }
            }
            else {
                return {
                    requestLists,
                    current: [...current, item]
                }
            }
        }), { current: [], requestLists: []})
    const batchPromises = [...groupBatches.requestLists, groupBatches.current]
        .filter((itemList) => (itemList.length))
        .map((itemList) => (dbClient.send(new BatchWriteItemCommand({ RequestItems: {
            [table]: itemList
        } }))))
    return Promise.all(batchPromises)
}

export const batchGetDispatcher = async ({ table, items, projectionExpression }) => {
    const groupBatches = items
        .reduce((({ current, requestLists }, item) => {
            if (current.length > 39) {
                return {
                    requestLists: [ ...requestLists, current ],
                    current: [item]
                }
            }
            else {
                return {
                    requestLists,
                    current: [...current, item]
                }
            }
        }), { current: [], requestLists: []})
    const batchPromises = [...groupBatches.requestLists, groupBatches.current]
        .filter((itemList) => (itemList.length))
        .map((itemList) => (dbClient.send(new BatchGetItemCommand({ RequestItems: {
            [table]: {
                Keys: itemList,
                ProjectionExpression: projectionExpression
            }
        } }))))
    const outcomes = await Promise.all(batchPromises)
    return outcomes.reduce((previous, { Responses = {} }) => {
        return [
            ...previous,
            ...(Responses[table] || []).map(unmarshall)
        ]}, [])
}

//
// mergeIntoDataRange queries a given data search (either using the primary sort if Asset/EphemeraId is
// specified in the search, or the DataCategory index if DataCategory is specified) and compares it
// against incoming data.  The 'table' argument should pass only 'assets' or 'ephemera' strings.
//
// Incoming records will either already possess the unmatched key (in the case where extractKey is null),
// or will generate that key by calling extractKey with two arguments:  first the element in question,
// and second the entire list of current items.  (e.g.:  This is used to match scoped keys to global
// IDs of Rooms in WML schemata)
//
// A given record _already_ in the database will either match (by Asset/EphemeraId and DataCategory) with
// an incoming record, or it will not.  Likewise, a given incoming record will either match a record
// already present, or it will not.
//
// The merge function takes a map of two-element tuples, incoming and current, indexed by the key
// (DataCategory, or Asset/EphemeraId) that is *not* specified in the search.  Each of these
// tuples will contain either:
//    * A match (both incoming and current elements specified)
//    * An unmatched incoming element
//    * An unmatched current element
//
// The merge function must return a different map on the same keys.  Each key will specify either:
//    * The string 'ignore', which will do nothing
//    * The string 'delete', which will eliminate the current element if present
//    * An object that, together with the two keys unique to that position in the map, will be
//      made into the body of a PutItem operation
//
export const mergeIntoDataRange = async ({
    table,
    search: {
        DataCategory,
        AssetId,
        EphemeraId
    },
    items,
    mergeFunction,
    extractKey = null
}) => {
    //
    // TODO:  Better error handling and validation throughout
    //
    const TableName = table === 'assets' ? assetsTable : ephemeraTable
    if ((DataCategory === undefined ? 0 : 1) + (AssetId === undefined ? 0 : 1) + (EphemeraId === undefined ? 0 : 1) > 1) {
        console.log(`ERROR: mergeIntoDataRange accepts only one of 'DataCategory', 'EphemeraId', and 'AssetId' search terms`)
        return;
    }
    const keyLabel = table === 'assets'
        ? AssetId ? 'DataCategory' : 'AssetId'
        : EphemeraId ? 'DataCategory': 'EphemeraId'
    const extractKeyDefault = (item) => (item[keyLabel])
    const KeyConditionExpression = DataCategory
        ? `DataCategory = :dc`
        : AssetId
            ? `AssetId = :aid`
            : `EphemeraId = eid`
    const { Items: dbItems = [] } = await dbClient.send(new QueryCommand({
        TableName,
        KeyConditionExpression,
        ExpressionAttributeValues: marshall({
            ":dc": DataCategory,
            ":aid": AssetId,
            ":eid": EphemeraId
        }, { removeUndefinedValues: true }),
        ...(DataCategory ? { IndexName: 'DataCategoryIndex' } : {})
    }))
    const currentItems = dbItems.map(unmarshall)
    const incomingItems = items.map((item) => ({
            [keyLabel]: extractKey ? extractKey(item, currentItems) : extractKeyDefault(item),
            ...item
        }))
    const firstPassMerging = currentItems.reduce((previous, item) => ({
            ...previous,
            [extractKeyDefault(item)]: { current: item }
        }), {})
    const secondPassMerging = incomingItems.reduce((previous, item) => {
        const key = extractKeyDefault(item)
        return {
            ...previous,
            [key]: {
                ...(previous[key] || {}),
                incoming: item
            }
        }
    }, firstPassMerging)
    const thirdPassMerging = Object.entries(secondPassMerging)
        .reduce((previous, [key, items]) => ({
            ...previous,
            [key]: mergeFunction(items)
        }), {})
    const deleteRecord = (key) => ({
        DeleteRequest: {
            Key: marshall({
                DataCategory,
                AssetId,
                EphemeraId,
                [keyLabel]: key
            }, { removeUndefinedValues: true })
        }
    })
    const fourthPassMerging = Object.entries(thirdPassMerging)
        .filter(([key, value]) => (value !== 'ignore'))
        .map(([key, value]) => {
            if (value === 'delete') {
                return deleteRecord(key)
            }
            else {
                return {
                    PutRequest: {
                        Item: marshall({
                            DataCategory,
                            AssetId,
                            EphemeraId,
                            [keyLabel]: key,
                            ...value
                        }, { removeUndefinedValues: true })
                    }
                }
            }
        })
    await batchWriteDispatcher({
        table: TableName,
        items: fourthPassMerging
    })
}

const abstractGetItem = (table) => async ({
    AssetId,
    EphemeraId,
    MessageId,
    DataCategory,
    ProjectionFields = [
        table === assetsTable
            ? 'AssetId'
            : table === ephemeraTable
                ? 'EphemeraId'
                : 'MessageId'
    ],
    ExpressionAttributeNames
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Item = {} } = await dbClient.send(new GetItemCommand({
            TableName: table,
            Key: marshall({
                AssetId,
                EphemeraId,
                MessageId,
                DataCategory
            }, { removeUndefinedValues: true }),
            ProjectionExpression: ProjectionFields.join(', '),
            ...(ExpressionAttributeNames ? { ExpressionAttributeNames } : {})
        }))
        return unmarshall(Item)
    })
}

export const assetGetItem = abstractGetItem(assetsTable)

export const assetQuery = abstractQuery(dbClient, assetsTable)

export const updateAsset = async ({
    AssetId,
    DataCategory,
    UpdateExpression,
    ExpressionAttributeNames,
    ExpressionAttributeValues
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new UpdateItemCommand({
            TableName: assetsTable,
            Key: marshall({
                AssetId,
                DataCategory
            }),
            UpdateExpression,
            ...(ExpressionAttributeValues ? { ExpressionAttributeValues: marshall(ExpressionAttributeValues, { removeUndefinedValues: true }) } : {}),
            ExpressionAttributeNames
        }))
    })
}

const abstractPutItem = (table) => async (item) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new PutItemCommand({
            TableName: table,
            Item: marshall(item, { removeUndefinedValues: true })
        }))
    })
}

//
// TODO:  Error handling to respond if the DynamoDB service throws an error
//
export const replaceItem = abstractPutItem(assetsTable)

export const putAsset = async ({
    Item
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new PutItemCommand({
            TableName: assetsTable,
            Item: marshall(Item, { removeUndefinedValues: true })
        }))
    })
}

export const deleteAsset = async ({
    AssetId,
    DataCategory
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new DeleteItemCommand({
            TableName: assetsTable,
            Key: marshall({
                AssetId,
                DataCategory
            })
        }))
    })
}

export const ephemeraGetItem = async ({
    EphemeraId,
    DataCategory,
    ProjectionFields = ['EphemeraId'],
    ExpressionAttributeNames,
    catchException = () => ({})
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Item = {} } = await dbClient.send(new GetItemCommand({
            TableName: ephemeraTable,
            Key: marshall({
                EphemeraId,
                DataCategory
            }),
            ProjectionExpression: ProjectionFields.join(', '),
            ...(ExpressionAttributeNames ? { ExpressionAttributeNames } : {})
        }))
        return unmarshall(Item)
    }, catchException)
}

export const ephemeraQuery = async ({
    EphemeraId,
    ProjectionFields = ['DataCategory'],
    ExpressionAttributeNames
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Items = [] } = await dbClient.send(new QueryCommand({
            TableName: ephemeraTable,
            KeyConditionExpression: 'EphemeraId = :ephemeraId',
            ExpressionAttributeValues: marshall({
                ':ephemeraId': EphemeraId
            }),
            ProjectionExpression: ProjectionFields.join(', '),
            ...(ExpressionAttributeNames ? { ExpressionAttributeNames } : {})
        }))
        return Items.map(unmarshall)
    }, () => ([]))
}

export const ephemeraDataCategoryQuery = async ({
    DataCategory,
    EphemeraPrefix,
    ProjectionFields = ['EphemeraId'],
    ExpressionAttributeNames
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Items = [] } = await dbClient.send(new QueryCommand({
            TableName: ephemeraTable,
            KeyConditionExpression: EphemeraPrefix ? 'DataCategory = :dc AND begins_with(EphemeraId, :EphemeraPrefix)' : 'DataCategory = :dc',
            IndexName: 'DataCategoryIndex',
            ExpressionAttributeValues: marshall({
                ':dc': DataCategory,
                ':EphemeraPrefix': EphemeraPrefix
            }, { removeUndefinedValues: true }),
            ProjectionExpression: ProjectionFields.join(', '),
            ...(ExpressionAttributeNames ? { ExpressionAttributeNames } : {})
        }))
        return Items.map(unmarshall)
    }, () => ([]))
}

export const ephemeraConnectionQuery = async ({
    ConnectionId,
    EphemeraPrefix,
    ProjectionFields = ['EphemeraId'],
    ExpressionAttributeNames
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Items = [] } = await dbClient.send(new QueryCommand({
            TableName: ephemeraTable,
            KeyConditionExpression: 'ConnectionId = :ConnectionId and begins_with(EphemeraId, :EphemeraPrefix)',
            IndexName: 'ConnectionIndex',
            ExpressionAttributeValues: marshall({
                ':ConnectionId': ConnectionId,
                ':EphemeraPrefix': EphemeraPrefix
            }),
            ProjectionExpression: ProjectionFields.join(', '),
            ...(ExpressionAttributeNames ? { ExpressionAttributeNames } : {})
        }))
        return Items.map(unmarshall)
    }, () => ([]))
}

export const putEphemera = async ({
    Item
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new PutItemCommand({
            TableName: ephemeraTable,
            Item: marshall(Item)
        }))
    })
}

export const deleteEphemera = async ({
    EphemeraId,
    DataCategory
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new DeleteItemCommand({
            TableName: ephemeraTable,
            Key: marshall({
                EphemeraId,
                DataCategory
            })
        }))
    })
}

export const updateEphemera = async ({
    EphemeraId,
    DataCategory,
    UpdateExpression,
    ExpressionAttributeNames,
    ExpressionAttributeValues,
    catchException = () => ({})
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Attributes = {} } = await dbClient.send(new UpdateItemCommand({
            TableName: ephemeraTable,
            Key: marshall({
                EphemeraId,
                DataCategory
            }),
            UpdateExpression,
            ...(ExpressionAttributeValues ? { ExpressionAttributeValues: marshall(ExpressionAttributeValues, { removeUndefinedValues: true }) } : {}),
            ExpressionAttributeNames
        }))
        return unmarshall(Attributes)
    }, catchException)
}

export const scanEphemera = async ({
    FilterExpression,
    ExpressionAttributeValues,
    ExpressionAttributeNames,
    ProjectionFields = ['EphemeraId'],
    ReturnValues
}) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new ScanCommand({
            TableName: ephemeraTable,
            FilterExpression,
            ExpressionAttributeValues: marshall(ExpressionAttributeValues, { removeUndefinedValues: true }),
            ExpressionAttributeNames,
            ProjectionExpression: ProjectionFields.join(', '),
            ReturnValues
        }))
    })
}

export const publishMessage = async (Item) => {
    return await asyncSuppressExceptions(async () => {
        await dbClient.send(new PutItemCommand({
            TableName: messageTable,
            Item: marshall({
                DataCategory: 'Meta::Message',
                ...Item
            })
        }))
    })
}

export const messageDataCategoryQuery = async ({
    DataCategory,
    ExpressionAttributeNames,
    ExclusiveStartKey
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Items = [], LastEvaluatedKey } = await dbClient.send(new QueryCommand({
            TableName: messageTable,
            KeyConditionExpression: 'DataCategory = :dc',
            IndexName: 'DataCategoryIndex',
            ExpressionAttributeValues: marshall({
                ':dc': DataCategory
            }),
            ExpressionAttributeNames,
            ExclusiveStartKey
        }))
        return {
            Items: Items.map(unmarshall),
            LastEvaluatedKey
        }
    }, () => ([]))
}

export const messageDeltaQuery = async ({
    Target,
    ExpressionAttributeNames,
    ExclusiveStartKey,
    StartingAt,
    Limit
}) => {
    return await asyncSuppressExceptions(async () => {
        const { Items = [], LastEvaluatedKey } = await dbClient.send(new QueryCommand({
            TableName: deltaTable,
            KeyConditionExpression: 'Target = :Target and DeltaId >= :Start',
            ExpressionAttributeValues: marshall({
                ':Target': Target,
                ':Start': `${StartingAt}`
            }),
            ExpressionAttributeNames,
            ExclusiveStartKey,
            Limit
        }))
        return {
            Items: Items.map(unmarshall),
            LastEvaluatedKey
        }
    }, () => ([]))
}