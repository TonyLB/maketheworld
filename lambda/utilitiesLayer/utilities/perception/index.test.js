import { jest, describe, expect, it } from '@jest/globals'

jest.mock('./dynamoDB.js')
import { getCharacterAssets, getRoomMeta, getStateByAsset } from './dynamoDB.js'

jest.mock('../dynamoDB/index.js')
import { ephemeraDB } from '../dynamoDB/index.js'

import { resultStateFactory, testMockImplementation } from '../executeCode/testAssets.js'

import { render } from './index.js'

describe('dependencyCascade', () => {

    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        jest.restoreAllMocks()
    })

    it('should return empty on an empty list', async () => {
        ephemeraDB.getItem.mockResolvedValue({
            assets: []
        })
        const output = await render([])
        expect(output).toEqual([])
    })

    it('should render with no provided state data', async () => {
        const testAssets = resultStateFactory()
        ephemeraDB.getItem.mockResolvedValue({
            assets: ['BASE', 'LayerA', 'LayerB', 'MixLayerA', 'MixLayerB']
        })
        getStateByAsset.mockImplementation(async (assets) => {
            return assets.reduce((previous, asset) => ({ ...previous, [asset]: testAssets[asset] || {} }), {})
        })
        getRoomMeta.mockResolvedValue({
            ['ROOM#MNO']: [
                {
                    DataCategory: 'ASSET#BASE',
                    appearances: [{
                        conditions: [{
                            if: 'foo',
                            dependencies: ['foo']
                        }],
                        render: ['Test One. '],
                        exits: []
                    }]
                }
            ]
        })
        getCharacterAssets.mockResolvedValue({ QRS: [] })
        const output = await render([{
            EphemeraId: 'ROOM#MNO',
            CharacterId: 'QRS'
        }])
        expect(output).toEqual([{
            Ancestry: '',
            CharacterId: 'QRS',
            Characters: [],
            Description: ['Test One. '],
            EphemeraId: 'ROOM#MNO',
            RoomId: 'MNO',
            Name: "",
            Exits: []
        }])
        expect(getStateByAsset).toHaveBeenCalledWith(['BASE'])
    })

    it('should render with provided state data', async () => {
        const testAssets = resultStateFactory()
        ephemeraDB.getItem.mockResolvedValue({
            assets: ['BASE', 'LayerA', 'LayerB', 'MixLayerA', 'MixLayerB']
        })
        getRoomMeta.mockResolvedValue({
            ['ROOM#MNO']: [
                {
                    DataCategory: 'ASSET#BASE',
                    appearances: [{
                        conditions: [{
                            if: 'foo',
                            dependencies: ['foo']
                        }],
                        render: ['Test One. '],
                        exits: []
                    }]
                }
            ]
        })
        getCharacterAssets.mockResolvedValue({ QRS: [] })
        const output = await render([{
            EphemeraId: 'ROOM#MNO',
            CharacterId: 'QRS'
        }], testAssets)
        expect(output).toEqual([{
            Ancestry: '',
            CharacterId: 'QRS',
            Characters: [],
            Description: ['Test One. '],
            EphemeraId: 'ROOM#MNO',
            RoomId: 'MNO',
            Name: "",
            Exits: []
        }])
        expect(getStateByAsset).toHaveBeenCalledWith([])

    })

    it('should render fetch data only where needed', async () => {
        const testAssets = resultStateFactory()
        ephemeraDB.getItem.mockResolvedValue({
            assets: ['BASE']
        })
        getStateByAsset.mockImplementation(async (assets) => {
            return assets.reduce((previous, asset) => ({ ...previous, [asset]: testAssets[asset] || {} }), {})
        })
        getRoomMeta.mockResolvedValue({
            ['ROOM#MNO']: [
                {
                    DataCategory: 'ASSET#BASE',
                    appearances: [{
                        conditions: [{
                            if: 'foo',
                            dependencies: ['foo']
                        }],
                        render: ['Test One. '],
                        exits: []
                    }]
                }
            ],
            ['ROOM#TUV']: [
                {
                    DataCategory: 'ASSET#LayerB',
                    appearances: [{
                        conditions: [{
                            if: 'baz',
                            dependencies: ['baz']
                        }],
                        render: ['Test Two. '],
                        exits: []
                    }]
                },
                {
                    DataCategory: 'ASSET#LayerA',
                    appearances: [{
                        conditions: [{
                            if: 'bar',
                            dependencies: ['bar']
                        }],
                        render: ['Test Three. '],
                        exits: []
                    }]
                }
            ]
        })
        getCharacterAssets.mockResolvedValue({ QRS: ['LayerB'], XYZ: ['LayerA'] })
        const output = await render([{
            EphemeraId: 'ROOM#MNO',
            CharacterId: 'XYZ'
        },
        {
            EphemeraId: 'ROOM#TUV',
            CharacterId: 'QRS'
        }], { BASE: testAssets.BASE })
        expect(output).toEqual([{
            Ancestry: '',
            CharacterId: 'XYZ',
            Characters: [],
            Description: ['Test One. '],
            EphemeraId: 'ROOM#MNO',
            RoomId: 'MNO',
            Name: "",
            Exits: []
        },
        {
            Ancestry: '',
            CharacterId: 'QRS',
            Characters: [],
            Description: ['Test Two. '],
            EphemeraId: 'ROOM#TUV',
            RoomId: 'TUV',
            Name: "",
            Exits: []
        }])

        expect(getStateByAsset).toHaveBeenCalledWith(['LayerB'])

    })

})