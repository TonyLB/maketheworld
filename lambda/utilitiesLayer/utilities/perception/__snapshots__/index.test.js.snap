// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should fetch assetList data only where needed 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "XYZ",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test One. ",
      },
    ],
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "MNO",
    "tag": "Room",
  },
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test Two. ",
      },
    ],
    "EphemeraId": "ROOM#TUV",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "TUV",
    "tag": "Room",
  },
]
`;

exports[`render should fetch state data only where needed 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "XYZ",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test One. ",
      },
    ],
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "MNO",
    "tag": "Room",
  },
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test Two. ",
      },
    ],
    "EphemeraId": "ROOM#TUV",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "TUV",
    "tag": "Room",
  },
]
`;

exports[`render should render maps 1`] = `
Array [
  Object {
    "CharacterId": "QRS",
    "EphemeraId": "MAP#MNO",
    "MapId": "MNO",
    "Name": "Grand Bazaar",
    "fileURL": "test.png",
    "rooms": Object {
      "BASE#fountainSquare": Object {
        "EphemeraId": "ROOM#XYZ",
        "exits": Array [
          Object {
            "name": "library",
            "to": "BASE#library",
          },
        ],
        "name": Array [
          "Fountain Square",
        ],
        "x": 0,
        "y": 100,
      },
      "BASE#library": Object {
        "EphemeraId": "ROOM#Library",
        "exits": undefined,
        "x": 0,
        "y": -100,
      },
    },
    "type": "Map",
  },
]
`;

exports[`render should render spacing around tags 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test ",
      },
      Object {
        "key": "testOne",
        "tag": "Link",
      },
      Object {
        "tag": "String",
        "value": " ",
      },
      Object {
        "key": "testTwo",
        "tag": "Link",
      },
      Object {
        "tag": "String",
        "value": "-attached",
      },
    ],
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "MNO",
    "tag": "Room",
  },
]
`;

exports[`render should render when mapValueOnly set true 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [
      Object {
        "Name": "vortex",
        "RoomId": "VORTEX",
        "Visibility": "Public",
      },
    ],
    "Name": "Test One. Test Two. ",
    "RoomId": "MNO",
    "tag": "Room",
  },
]
`;

exports[`render should render with no provided state data 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test One. ",
      },
    ],
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "MNO",
    "tag": "Room",
  },
]
`;

exports[`render should render with provided state data 1`] = `
Array [
  Object {
    "Ancestry": "",
    "CharacterId": "QRS",
    "Characters": Array [],
    "Description": Array [
      Object {
        "tag": "String",
        "value": "Test One. ",
      },
    ],
    "EphemeraId": "ROOM#MNO",
    "Exits": Array [],
    "Features": Array [],
    "Name": "",
    "RoomId": "MNO",
    "tag": "Room",
  },
]
`;
